### PR TITLE
[IMP] fields: forbid required m2o ondelete set null

### PR DESCRIPTION
**Pull request by Elkasitu**
A many2one field that is required means that its column cannot contain
null values, however the default ondelete policy for all m2o fields is
`set null`, which contradicts `required=True`.

This patch will, by default, apply the `restrict` policy to required m2o
fields unless the ondelete attribute is explicitly specified. However,
if the policy specified is `set null`, a ValueError will be raised
whenever the registry registers the field as this makes no sense.

The default ondelete policy remains unchanged for non-required m2o
fields (set null)

### CODE REVIEW

File: odoo/addons/test_new_api/tests/test_new_fields.py [hunk 1]
```diff
@@ -1435,3 +1435,28 @@ def test_move_N_cycle(self):
...
             self.cats(1, 3).write({'parent': self.cats(9).id})
+
+
+class TestRequiredMany2one(common.TransactionCase):
+
+    def test_explicit_ondelete(self):
+        f = self.env['ir.model.fields'].search([
+            ('model', '=', 'test_new_api.req_m2o'),
+            ('name', '=', 'foo'),
+        ])
```
Comment (rco-odoo) [inline odoo/addons/test_new_api/tests/test_new_fields.py]: There's an API for that:
```python
f = self.env['ir.model.fields']._get('test_new_api.req_m2o', 'foo')
```

File: odoo/addons/test_new_api/tests/test_new_fields.py [hunk 2]
```diff
@@ -1435,3 +1435,28 @@ def test_move_N_cycle(self):
...
+        f = self.env['ir.model.fields'].search([
+            ('model', '=', 'test_new_api.req_m2o'),
+            ('name', '=', 'bar'),
+        ])
+        self.assertEqual(f.on_delete, 'restrict')
+
+    def test_explicit_set_null(self):
+        ReqM2o = self.env['test_new_api.req_m2o']
+        f = ReqM2o._fields['foo']
+        f.ondelete = 'set null'
```
Comment (rco-odoo) [inline odoo/addons/test_new_api/tests/test_new_fields.py]: This is not undone after the test. Please use `self.patch()`!

File: odoo/fields.py [hunk 3]
```diff
@@ -2068,6 +2068,31 @@ def _setup_attrs(self, model, name):
...
+            if self.ondelete == 'set null':
+                raise ValueError(
+                    "The m2o field %s of model %s is required but declares its ondelete policy "
+                    "as being 'set null'. Only 'restrict' and 'cascade' make sense."
+                    % (self.name, model._name)
+                )
+            elif not self.ondelete:
+                self.ondelete = 'restrict'
+        elif not self.required and not self.ondelete:
+            self.ondelete = 'set null'
```
Comment (rco-odoo) [inline odoo/fields.py]: The logic is more complicated than necessary to me.
```python
if not self.ondelete:
    # provide sensible default
    self.ondelete = 'restrict' if self.required else 'set null'
elif self.ondelete == 'set null' and self.required:
    raise ValueError(...)
```

File: odoo/addons/test_new_api/tests/test_new_fields.py [hunk 4]
```diff
@@ -1435,3 +1435,22 @@ def test_move_N_cycle(self):
...
+        f = self.env['ir.model.fields']._get('test_new_api.req_m2o', 'foo')
+        self.assertEqual(f.on_delete, 'cascade')
+
+    def test_implicit_ondelete(self):
+        f = self.env['ir.model.fields']._get('test_new_api.req_m2o', 'bar')
+        self.assertEqual(f.on_delete, 'restrict')
+
+    def test_explicit_set_null(self):
+        ReqM2o = self.env['test_new_api.req_m2o']
+        f = ReqM2o._fields['foo']
```
Comment (Elkasitu) [inline odoo/addons/test_new_api/tests/test_new_fields.py]:
```python3
self.addCleanup(self.env.registry.setup_models, self.cr)
```

File: odoo/fields.py [hunk 5]
```diff
@@ -2068,6 +2068,28 @@ def _setup_attrs(self, model, name):
...
+        #      specified, in which case it is left as-is.
+        #   3. The m2o field is required and its ondelete attribute has not been explicitly
+        #      specified, in which case it defaults to `restrict`.
+        #   4. The m2o field is required and its ondelete attribute has been explicitly specified
+        #      as either `restrict` or `cascade`, in which case it is left as-is.
+        #   5. The m2o field is required and its ondelete attribute has been explicitly specified
+        #      as `set null`, this raises a ValueError as it's considered a programmer error.
+        if not self.ondelete:
+            self.ondelete = 'restrict' if self.required else 'set null'
+        elif self.ondelete == 'set null' and self.required:
```
Comment (Elkasitu) [inline odoo/fields.py]: `if`

File: odoo/addons/test_new_api/tests/test_new_fields.py [hunk 6]
```diff
@@ -1435,3 +1435,22 @@ def test_move_N_cycle(self):
         """ Move multiple nodes to create a cycle. """
         with self.assertRaises(UserError):
             self.cats(1, 3).write({'parent': self.cats(9).id})
+
+
+class TestRequiredMany2one(common.TransactionCase):
+
+    def test_explicit_ondelete(self):
+        f = self.env['ir.model.fields']._get('test_new_api.req_m2o', 'foo')
```
Comment (rco-odoo) [inline odoo/addons/test_new_api/tests/test_new_fields.py]: Why don't you check the field on the model instead?
```suggestion
        foo = self.env['test_new_api.req_m2o']._fields['foo']
```
This makes the test independent from the reflection of the field.

File: odoo/fields.py [hunk 7]
```diff
@@ -2068,6 +2068,28 @@ def _setup_attrs(self, model, name):
...
+        #   1. The m2o field is not required and its ondelete attribute has not been explicitly
+        #      specified, in which case it defaults to `set null`.
+        #   2. The m2o field is not required and its ondelete attribute has been explicitly
+        #      specified, in which case it is left as-is.
+        #   3. The m2o field is required and its ondelete attribute has not been explicitly
+        #      specified, in which case it defaults to `restrict`.
+        #   4. The m2o field is required and its ondelete attribute has been explicitly specified
+        #      as either `restrict` or `cascade`, in which case it is left as-is.
+        #   5. The m2o field is required and its ondelete attribute has been explicitly specified
+        #      as `set null`, this raises a ValueError as it's considered a programmer error.
```
Comment (rco-odoo) [inline odoo/fields.py]: I would mention three cases:
1. The ondelete attribute is not set; provide a consistent default value.
2. The ondelete attribute is set to 'set null' and the field is required; this case is considered a programming error.
3. The ondelete attribute is set and consistent; nothing is done.

**Comment by rco-odoo** [comment]
@robodoo r+

**Comment by Elkasitu** [comment]
@robodoo retry

**Comment by Elkasitu** [comment]
@robodoo retry REEEEEEEEEEEEEE

**Pull request closed without merging**